pkg/apis/core/types.go

```diff
@@ -1670,6 +1673,29 @@ type CSIVolumeSource struct {
 	NodePublishSecretRef *LocalObjectReference
 }
 
+// EphemeralVolumeSource represents ephemeral storage that is handled by a normal storage driver.
+type EphemeralVolumeSource struct {
+	// Will be created as a stand-alone PVC to provision the volume.
+	//
+	// The pod in which this EphemeralVolumeSource is embedded will be the
+	// owner, i.e. the PVC will be deleted together with the pod.
+	// The name of the PVC will be `<pod name>-<volume name>` where `<volume name>`
+	// is the name from the `PodSpec.Volumes` array entry. Care must
+	// be taken to keep this concatendated name shorter enough that it
+	// can be used as object name. Labels will be copied from the pod.
+	//
+	// This field is read-only and no changes will be made by Kubernetes
+	// to the PVC after it has been created.
+	//
+	// Required, must not be nil.
+	VolumeClaim *PersistentVolumeClaimSpec
```

Comment:
A generic ephemeral volume.... is not persistent.  Why "ephemeral" pointing to something called "persistent"?

Reply:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1698-generic-ephemeral-volumes#motivation

I really don't understand this motivation in the KEP in terms of helping me understand what this PR is supposed to change (as a reviewer, consumer of kube, and general critic of things that get added).

The motivation sounds like "we want things that are not ephemeral to be shown as ephemeral", but I don't know *why*.  Can you help me unpack what the intent of that KEP is in more controlled way (so we can update the KEP) and I can frame this PR in terms of what the goal is?

Reply:
The goal is to support easy deployment of apps that need storage
- that is provided by a CSI driver and
- only as long as a pod runs.

The KEP mentions adding local storage to applications like memcached as use case for this. Here storage serves as an extension or replacement for DRAM. Persistency is not needed and just makes deployment harder.

None of the existing mechanisms fit that goal:
- EmptyDir isn't flexible enough (cannot be provided through a CSI driver).
- Deployment, daemonset, replicas all don't support per-pod volumes. Statefulset is the only one which does, but doesn't delete volumes, i.e. additional logic on top of it is needed to remove them, and it might not have the right semantic for the app (for example, the KubeDB memcached operator uses a deployment).
- CSI ephemeral inline volumes don't support storage capacity tracking and pod startup can fail badly when pods land on a node with insufficient storage; this is by design and would be very hard to fix.

Regarding the naming: if you replace "PersistentVolumeClaim" with "VolumeClaim", the proposal makes sense again. The name is basically historic and just cannot be changed anymore. The old name implies a semantic that just isn't there: the volume always was meant to persist only as long as the corresponding claim (ignore retention policy for the sake of simplicity), and that still holds here. The only difference is that there is a predetermined trigger for deleting the PVC.

Here's an example of deploying memcached with a CSI ephemeral inline volume:
https://github.com/intel/pmem-csi/blob/devel/examples/memcached.md => https://github.com/intel/pmem-csi/blob/devel/deploy/kustomize/memcached/ephemeral/memcached-ephemeral.yaml

CSI ephemeral inline volumes kind of work, but only because in PMEM-CSI a scheduler extender tries to mitigate the risk of bad scheduling decisions. It cannot prevent them entirely and is hard to enable in a cluster. Other CSI  drivers don't have such a Kubernetes-specific extension.

Reply:
@smarterclayton, maybe it could be emphasized better in the KEP, but the overall motivation distilled into two words is "better emptyDir" - when the root disk is too slow / too small / too shared for pod's ephemeral data.

Yes, using "PersistentVolumeClaim" for something that's called "ephemeral" is not great. We spent quite some time with this in sig-storage and generic ephemeral volume was the best candidate. Suggestions for better naming are welcome.

It has also some other use cases (you can populate it with a restored snapshot / clone a volume instead), still, it's just a scratch space with emptyDir lifetime.